### DISCUSSION TITLE
gRPC Uploading and Downloading small files/blobs.

### DISCUSSION THREAD

**Discussion by enzopsc** [seaweedfs/seaweedfs#2297] [General]
Hi. I was developing a gRPC client for all 3 interfaces ( Master, Volume and Filer ). Unfortunately I wasn't able to Upload or Download FIles/blobs through it. Is there any way to make it work through gRPC? I was kinda able to write on volumes with follow steps:
1) Master.Assign - To get a fileId
2) Volume.ReadVolumeFileStatus - To get current dat_file_size ( Starting offset during file/blob read)
3) Building a compatible Volume's blob
4) Volume.WriteNeedleBlob - To append blob
5) Volume.ReadVolumeFileStatus - To get current dat_file_size ( Ending offset during file/blob append )
With that I was able to append blobs properly to volumes but I'm kinda confused.
So I was wondering few things:
What I did is correct? Is there any better way to work with gRPC to upload|download blob?
Why do ReadNeedleBlob accept needle_id as parameter but doesn't use it to get blob? Isn't it supposed to return the blob relative to the needle_id?
Thanks.

**Comment by chrislusf** [comment]
Seems OK. Need to take care of replications.

`ReadNeedleBlob` was used by command_volume_check_disk.go only. It was not intended as a general API. We can add some logic to read by needle_id when offset and size are zero.

**Reply by enzopsc** [reply]
Well I thought that gRPC was also made for that since it has Read/Write operations, so if you can add this logic would be really appreciated. 
But I didn’t get about replications. Do you mean that I should manually replicate blobs across volumes?

**Reply by enzopsc** [reply]
Well after looking at APIs I will just stick with Filer's API. Thanks.

**Reply by chrislusf** [reply]
> But I didn’t get about replications. Do you mean that I should manually replicate blobs across volumes?

Correct. The API only writes to one volume server.
